VisualPinball.Unity/VisualPinball.Unity.Editor/Editors/DragPointsEditor.cs

```diff
@@ -2,72 +2,465 @@
 using UnityEditor;
 using UnityEngine;
 using VisualPinball.Engine.Math;
+using VisualPinball.Unity.Editor.Inspectors;
 using VisualPinball.Unity.Extensions;
 using VisualPinball.Unity.VPT;
 
 namespace VisualPinball.Unity.Editor.Editors
 {
-	public class DragPointsEditor
+	public class DragPointsEditor 
 	{
-		private List<DragPointData> _controlPoints = new List<DragPointData>();
+		public DragPointsEditor(ItemInspector _inspector) { _itemInspector = _inspector; }
+
+		public class ControlPoint
+		{
+			static public float ScreenRadius = 0.25f;
+			public DragPointData DragPoint;
+			public Vector3 WorldPos = Vector3.zero;
+			public Vector3 ScrPos = Vector3.zero;
+			public bool IsSelected = false;
+			public bool IsFoldOut = false;
+			public int ControlId = 0;
+
+			public ControlPoint(ref DragPointData dp, int controlID)
+			{
+				DragPoint = dp;
+				ControlId = controlID;
+			}
+		}
+
+		private Object _target = null;
+		private ItemInspector _itemInspector = null;
+
+		//Control points storing & rendering
+		private List<ControlPoint> _controlPoints = new List<ControlPoint>();
 		private List<Vector3> _pathPoints = new List<Vector3>();
 
+		//Control points position Handle
+		private List<ControlPoint> _selectedCP = new List<ControlPoint>();
+		private int _positionHandleControlId = 0;
+		private Vector3 _positionHandlePosition = Vector3.zero;
+
+		//Curve Traveller 
+		static public float CurveTravellerSizeRatio = 0.75f;
+		private int _curveTravellerControlId = 0;
+		private Vector3 _curveTravellerPosition = Vector3.zero;
+		private bool _curveTravellerVisible = false;
+		private int _curveTravellerControlPointIdx = -1;
+
+		//Inspector
+		private bool _foldoutControlPoints = false;
+		private GUIStyle _styleSelected = new GUIStyle();
+		private GUIStyle _styleLocked = new GUIStyle();
+
+		//Mesh management & Undo
+		private bool _needMeshRebuilt = false;
+		private string _pendingUndoMessage = string.Empty;
+
+		//Drop down PopupMenus
+		class MenuItems
+		{
+			public const string CONTROLPOINTS_MENUPATH = "CONTEXT/DragPointsEditor/ControlPoint";
+			public const string CURVETRAVELLER_MENUPATH = "CONTEXT/DragPointsEditor/CurveTraveller";
+
+			private static DragPointData RetrieveDragPoint(DragPointsEditor editor, int controlId)
+			{
+				if (editor == null)
+					return null;
+				return editor.GetDragPoint(controlId);
+			}
+
+			//Drag Points
+			[MenuItem(CONTROLPOINTS_MENUPATH + "/IsLocked", false, 1)]
+			private static void Lock(MenuCommand command)
+			{
+				ItemInspector editor = command.context as ItemInspector;
+				if (editor == null || editor.DragPointsEditor == null)

```

Comment:
Out of interest, how can this happen?

Reply:
Shouldn't...but you'll never know with asynchronous popmenu calls. Well...maybe overreacting :D